keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -27,79 +29,84 @@ class EnvironmentPromotionChecker(
+                    .toMutableSet()
+
                   /**
-                   * We don't need to re-invoke stateful constraint evaluators for these, but we still
-                   * check stateless constraints to avoid approval outside of allowed-times.
+                   * Approve all constraints starting with oldest first so that the ordering is
+                   * maintained.
                    */
-                  log.debug(
-                    "Version $v of artifact ${artifact.name} is queued for approval, " +
-                      "and being evaluated for stateless constraints in environment ${environment.name}"
-                  )
-                  if (constraintRunner.checkStatelessConstraints(artifact, deliveryConfig, v, environment)) {
-                    approveVersion(deliveryConfig, artifact, v, environment.name)
-                    repository.deleteQueuedConstraintApproval(deliveryConfig.name, environment.name, v, artifact.reference)
+                  queuedForApproval
+                    .sortedWith(artifact.versioningStrategy.comparator.reversed())
+                    .forEach { v ->
+                      /**
+                       * We don't need to re-invoke stateful constraint evaluators for these, but we still
+                       * check stateless constraints to avoid approval outside of allowed-times.
+                       */
+                      log.debug(
+                        "Version $v of artifact ${artifact.name} is queued for approval, " +
+                          "and being evaluated for stateless constraints in environment ${environment.name}"
+                      )
+                      if (constraintRunner.checkStatelessConstraints(artifact, deliveryConfig, v, environment)) {
+                        approveVersion(deliveryConfig, artifact, v, environment.name)
+                        repository.deleteQueuedConstraintApproval(deliveryConfig.name, environment.name, v, artifact.reference)
+                      }
+                    }
+
+                  val versionSelected = queuedForApproval
+                    .sortedWith(artifact.versioningStrategy.comparator.reversed())
+                    .lastOrNull()
+                  if (versionSelected == null) {
+                    log.warn("No version of {} passes constraints for environment {}", artifact.name, environment.name)
                   }
                 }
-
-              val versionSelected = queuedForApproval
-                .sortedWith(artifact.versioningStrategy.comparator.reversed())
-                .lastOrNull()
-              if (versionSelected == null) {
-                log.warn("No version of {} passes constraints for environment {}", artifact.name, environment.name)
+              } else {
+                log.warn("Skipping checks for {} as it is not used in environment {}", artifact.name, environment.name)
```

Comment:
Is this really warn level? Seems more like debug

Reply:
Fair